fenix/app/src/main/java/org/mozilla/fenix/onboarding/view/JunoOnboardingMapper.kt

```diff
@@ -13,24 +17,83 @@ import org.mozilla.fenix.settings.SupportUtils
  * Returns a list of all the required Nimbus 'cards' that have been converted to [OnboardingPageUiData].
  */
 internal fun Collection<OnboardingCardData>.toPageUiData(
+    context: Context,
     showNotificationPage: Boolean,
     showAddWidgetPage: Boolean,
+    conditions: Map<String, String>,
 ): List<OnboardingPageUiData> =
     filter {
-        when (it.cardType) {
+        val typeFilter = when (it.cardType) {
             OnboardingCardType.NOTIFICATION_PERMISSION -> {
                 it.enabled && showNotificationPage
             }
+
             OnboardingCardType.ADD_SEARCH_WIDGET -> {
                 it.enabled && showAddWidgetPage
             }
+
             else -> {
                 it.enabled
             }
         }
+
+        if (!typeFilter) {
+            false
+        } else {
+            it.shouldDisplayCard(context, conditions)
+        }
     }.sortedBy { it.ordering }
         .map { it.toPageUiData() }
 
+private fun OnboardingCardData.shouldDisplayCard(
+    context: Context,
+    conditionsTable: Map<String, String>,
+): Boolean {
+    val jexlCache: MutableMap<String, Boolean> = mutableMapOf()
+    val helper = context.components.analytics.messagingStorage.helper
+
+    return verifyConditionEligibility(
+        prerequisites,
+        conditionsTable,
+        jexlCache,
+        helper,
+    ) && !verifyConditionEligibility(disqualifiers, conditionsTable, jexlCache, helper)
+}
+
+private fun verifyConditionEligibility(
+    cardConditions: List<String>,
+    conditionsTable: Map<String, String>,
+    jexlCache: MutableMap<String, Boolean>,
+    helper: GleanPlumbMessageHelper,
+): Boolean {
+    val malFormedMap = mutableMapOf<String, String>()
+    // Make sure conditions exist and have a value, and that the number
+    // of valid conditions matches the number of conditions on the card's
+    // respective prerequisite or disqualifier table. If these mismatch,
+    // that means a card contains a condition that's not in the feature
+    // conditions lookup table. JEXLs can only be evaluated on
+    // supported conditions. Otherwise, consider the card invalid.
+    val conditions = cardConditions.mapNotNull { conditionsTable[it] }
+    return if (conditions.size == cardConditions.size) {
+        return conditions.all { condition ->
+            jexlCache[condition]
+                ?: try {
+                    if (malFormedMap.containsKey(condition)) {
+                        return false
+                    }
+                    helper.evalJexl(condition).also { result ->
```

Comment:
I think something is not quite right here as `condition` here will be `true` or `false` which is then stored against `true` or `false`?

Reply:
should ` conditions.all` be `cardConditions.all`?